packages/nimble-components/src/wafer-map/specs/features/rendering.md

```diff
@@ -0,0 +1,207 @@
+
+[Worker Rendering POC](https://github.com/ni/nimble/compare/main...users/munteannatan/scale_and_test_updates)
+
+## Implementation / Design
+
+We will change the whole process of ingesting the data and rendering it in the canvas. The main changes are the data structures used and the process from single threaded to multi threaded. We expect that using better structured data will lead to time improvements when iterating over values which will improve the rendering and allow distributing the workload if necessary. Moreover, the adoption of thread based processing will free the main thread of the browser to handle user requests and other page events, while allowing us to scale in the future by parallel processing if the need arises.
+
+The detailed plan can be found in the azure HLD [Faster Rendering of Large Wafer Maps HLD](https://ni.visualstudio.com/DevCentral/_git/Skyline?path=/docs/design-documents/Ozone/Requirements/Optimize-Wafer-Map-to-Handle-NXP-s-340k-die-per-wafer-data-set/Faster-rendering-of-Large-Wafer-Maps.md&version=GC00f8bb9e698a7310f68fa54395eee63a99cec368&_a=preview).
+
+The POC is found in this branch [Worker Rendering POC](https://github.com/ni/nimble/compare/main...users/munteannatan/scale_and_test_updates).
+
+### Data Structure and Interface
+
+This is the proposed representation of the wafer information in memory
+
+```TS
+class WaferData {
+    // the x coordinates of each column of dies
+    dieColIndexArray: Int32Array;
+    // the lengths of each row of dies
+    rowLengthsArray: Int32Array;
+    // the y coordinates of each die as a matrix row by row
+    dieRowIndexLayer: Int32Array;
+    // the value of each die as a matrix row by row
+    dieValuesLayer: Int32Array;
+    // the highlight state of each die as a matrix row by row
+    dieHighlightsLayer: Int8Array;
+    // extra values layers
+    extraLayers : {
+        name: string;
+        type: string;
+        buffer: TypedArray;
+    }[]
+}
+```
+
+Using TypedArrays has the benefit of direct transfer to web workers without structured cloning of the object by transferring the arrayBuffers and reconstructing the object.
+
+Other benefits of typedArrays include the low access time when iterating over the values, more memory efficiency and faster direct access to metadata layers values.
+
+The previous inputs can be adapted to this new structure to maintain backwards compatibility.
+
+This API will have [optimized byte-array interop from Blazor](https://learn.microsoft.com/en-us/dotnet/core/compatibility/aspnet-core/6.0/byte-array-interop) and should be supported by Angular as a [vanilla javascript feature](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer).
+
```

Comment:
In this section we should list the alternative of Apache Arrow and the conditions for it:

Pros:

- A row based format that aligns well with the existing public api
- Well supported and tested format
- Nice public API to use, don't have to invent a new format, just doocument our schema for the arrow tables
- Designed for large dataset visualizations

What we need to prototype:

- Does it have comparable memory performance
- Does it perform well or have significant overhead
- What other criteria are y'all interested in in order to evaluate usage?